bot/utils/database-migrations.js

```diff
@@ -0,0 +1,88 @@
+const fs = require("fs");
+const db = require("../apis/database-crud.js");
+const Logger = require("./logger.js");
+
+/**
+ * Get a full list of available migrations, ordered from oldest to newest version.
+ * @returns {Array<string>} List of migrations, formatted like yyyy-mm-dd-name.js
+ */
+module.exports.listMigrations = () => {
+    //todo: is this ordering reliable on all OSes and all configurations?
+    return fs.readdirSync(__dirname + "/../migrations/");
+};
+
+/**
+ * Gets the latest migration version.
+ * @returns {string} Full name of latest migration version.
+ */
+module.exports.latestMigration = () => {
+    const list = this.listMigrations();
+    return list[list.length - 1];
+};
+
+/**
+ * Gets the currently applied migration.
+ * @returns {Promise<string>} Promise resolving to the current migration's full name.
+ */
+module.exports.current = () => {
+    return db.select("_Migrations", ["value"], {
+        name: "version"
+    }).then(results => {
+        if (results.length === 0) {
+            //Table does not contain a version key-value
+            return null;
+        } else {
+            return results[0].value;
+        }
+    }).catch((err) => {
+        if (err.errno === 1) {
+            //Assume "no such table" error (has no more specific code than that)
+            //todo: can we rely on the error message -> filter string
+            return null;
+        } else {
+            Logger.err(err);
+        }
+    });
+};
+
+/**
+ * Gets the differences required to get from one version to another.
+ * @param {string|null} oldVersion Full name of old version. If null, "no" version is used (index -1).
+ * @param {string|null} newVersion Full name of new version. If null, latest version is used.
+ * @returns {Array<Array<string>>} List of diff data, structured like [migrationName, "up"|"down"]
+ */
+module.exports.listDiff = (oldVersion, newVersion) => {
+    if (oldVersion === undefined || newVersion === undefined) {
+        throw new Error("oldVersion and newVersion may not be undefined.");
+    }
+
+    const migrations = this.listMigrations();
+
+    //Figure out the indices
+    let oldIndex;
+    if (oldVersion === null) {
```

Comment:
I think it should not be possible for `migrations` to contain the null value, so we can just use indexOf (because indexOf(null) will return -1 as long as null is not in the array)

Reply:
You are right that it is possible to simplify the code like that. But -1 is a special value that should only come for null ("before the first version"), it might seem like an oversight.

You did make me notice that, if a non-existing version is passed, -1 is also used which can cause a lot of problems, so I will add an exception for that case.

Reply:
Hmm... So what *should* happen when a version is not found?

Let's say we have v1 v2 v3, we are on v3 right now. We revert a commit so now only v1 v2 js files exist in the folder. If we run Discotron, it can't downgrade because it can't find v3.js.

I added an exception in the latest commit. But should we simply continue program execution? We can't be sure that the database version didn't for example rename a column or something, breaking the old code... How would you recover from that state?